### DISCUSSION TITLE
可能的死锁

### DISCUSSION THREAD

**Discussion by warren830** [panjf2000/ants#221] [Ideas]
不知道大家在使用的时候有没有遇到submit里面再在当前pool中再submit的情况，这种情况会进而引发死锁，目前我是用一个sub pool来解决的，不知道大家遇到的时候是怎么做的？

**Comment by panjf2000** [comment]
有没有实际代码可以复现？

**Reply by warren830** [reply]
我下面贴出了相关的简化代码

**Comment by warren830** [comment]
```go
func (apiClient *ApiAsyncClient) DoAsync(retry int) error {
   return apiClient.scheduler.Submit(func() error {
      _, err := apiClient.Do()
      if err != nil {
         if retry < apiClient.maxRetry {
            return apiClient.DoAsync(retry+1)
         }
      }
      return err
   })
}
```

**Comment by panjf2000** [comment]
你这不是死锁吧，只是把 pool 用完了，自然就阻塞住了，如果不想 pool 用完之后阻塞住就设置一下非阻塞 submit，还有下次用 markdown 语法格式化一下代码。

**Comment by warren830** [comment]
对，我的理解就是pool用完了导致的死锁，目前我们是增加了一个subpool来处理
________________________________
发件人: Andy Pan ***@***.***>
发送时间: 2022年5月5日 17:49
收件人: panjf2000/ants ***@***.***>
抄送: Yingchu Chen ***@***.***>; Author ***@***.***>
主题: Re: [panjf2000/ants] 可能的死锁 (Discussion #221)


你这不是死锁吧，只是把 pool 用完了，自然就阻塞住了，如果不想 pool 用完之后阻塞住就设置一下非阻塞 submit，还有下次用 markdown 语法格式化一下代码。

―
